qiskit_experiments/data_processing/data_processor.py

```diff
@@ -101,31 +112,56 @@ def _call_internal(
             history_nodes: The nodes, specified by index in the data processing chain, to
                 include in the history. If None is given then all nodes will be included
                 in the history.
+            call_up_to_node: The data processor will use each node in the processing chain
+                up to the node indexed by call_up_to_node. If this variable is not specified
+                then all nodes in the data processing chain will be called.
+            options: Run-time options given as keyword arguments that will be passed to the nodes.
 
         Returns:
             datum_ and history if with_history is True or datum_ if with_history is False.
 
         Raises:
             DataProcessorError: If the input key of the data processor is not contained in datum.
         """
+        if call_up_to_node is None:
+            call_up_to_node = len(self._nodes)
 
         if self._input_key not in datum:
             raise DataProcessorError(
                 f"The input key {self._input_key} was not found in the input datum."
             )
 
         datum_ = datum[self._input_key]
+        error_ = None
 
         history = []
         for index, node in enumerate(self._nodes):
-            datum_ = node(datum_)
 
-            if with_history and (
-                history_nodes is None or (history_nodes and index in history_nodes)
-            ):
-                history.append((node.__class__.__name__, datum_, index))
+            if index < call_up_to_node:
+                datum_, error_ = node(datum_, error_, **options)
+
+                if with_history and (
+                    history_nodes is None or (history_nodes and index in history_nodes)
+                ):
+                    history.append((node.__class__.__name__, datum_, error_, index))
 
         if with_history:
-            return datum_, history
+            return datum_, error_, history
         else:
-            return datum_
+            return datum_, error_
+
+    def train(self, data: List[Dict[str, Any]]):
```

Comment:
Shoudn't training be something that produces a DataProcessor from some training set, rather than a property of the data processor itself?

Reply:
We need to support the case where a data processor self-trains on the data. See the comment below on spectroscopy.